### DISCUSSION TITLE
API Page Pattern

### DISCUSSION THREAD

**Discussion by waldo1001** [microsoft/alguidelines#60] [BC Patterns]
We need a description of the "minimum" and the "best practices" of setting up an API.

Things I'm thinking of:
It should contain at least these fields
- id - SystemId
- lastModifiedDateTime - SystemModifiedAt
- Extensible = false;

Guidelines around the properties:
- APIPublisher
- APIGroup
- APIVersion
- EntityName
- EntitySetName
- ODataKeyFields = systemId

Refer to a snippet?
tpageapiwaldo is a start, but is far from perfect.

Is it
- A Pattern?
- A Guideline

**Comment by pibcht** [comment]
> lastModifiedDateTime - SystemModifiedAt

For this point, I've several thoughts:
- Why `lastModifiedDateTime` and not `modifiedAt` as this is a system field?
- Why not `createdDateTime` (Or better: `createdAt`) ? I'm not comfortable with "Modified" date for some immutable data

My version of that would rather be:

> It should contain at least these fields
> - id - SystemId
> - modifiedAt - SystemModifiedAt (except for `Editable = false`)
> - createdAt - SystemCreatedAt
> - Extensible = false;

And you can object to the idea of ​​having an exception in the "at least" rule 😅

**Comment by waldo1001** [comment]
@ajkauffmann is going to take the lead on this one...

**Comment by waldo1001** [comment]
Important: the name "lastModifiedDateTime" for webhook behavior:

![image](https://user-images.githubusercontent.com/12088142/153201393-8491cb7e-14fc-441b-9cc6-cc20689f1c3c.png)
